### PR TITLE
feat(table-semantic): expose col, colIndex, row, and rowIndex props

### LINKED ISSUE #3208: [TableSemantic] Expose row and rowIndex to TableBuilder components
Someone recently requested the ability to highlight a specific row with a background color. This isn't currently possible as we don't expose the current row or rowIndex as properties of the table row and cell styled components.

I suggest we add `$row` and `$rowIndex` properties to all row and cell components and that we add `$col` and `$colIndex` properties to all cell components so that this information is available to the function form of style overrides.

Example use:

```jsx
export default () => {
  const [selectedRow, setSelectedRow] = useState(0);

  return (
    <TableBuilder data={DATA} overrides={{
      TableBodyRow: {
        style: ({ $theme, $rowIndex }) => {
          if ($rowIndex === selectedRow) {
            return { backgroundColor: $theme.colors.warning600 };
          }
        },
      }
    }}>
      <TableBuilderColumn header="Produce">
        {(row, index) => (
          <div onClick={() => setSelectedRow(index)}>{row.bar}</div>
        )}
      </TableBuilderColumn>
      <TableBuilderColumn header="Quantity" numeric>
        {(row, index) => (
          <div onClick={() => setSelectedRow(index)}>{row.foo}</div>
        )}
      </TableBuilderColumn>
    </TableBuilder>
  );
};
```

### PR DESCRIPTION
**Pull request by shannonmoeller**
Fixes #3208 

#### Description

Exposes row and column information to styled components as `$col`, `$colIndex`, `$row`, `$rowIndex` properties. This will allow overrides access to this information in order to make style changes like highlighting a row for example.

#### Scope

- [ ] Patch: Bug Fix
- [x] Minor: New Feature
- [ ] Major: Breaking Change

### CODE REVIEW

File: src/table-semantic/styled-components.js [hunk 1]
```diff
@@ -48,59 +48,72 @@ export const StyledTableHeadRow = styled<{}>('tr', ({$theme}) => {
...
-      position: 'absolute',
-      top: '0',
-      right: '100%',
-      bottom: '0',
-      borderLeftColor: $theme.borders.border300.borderColor,
-      borderLeftStyle: $theme.borders.border300.borderStyle,
-      borderLeftWidth: $theme.borders.border300.borderWidth,
-    },
+type StyledTableHeadCellPropsT = {
+  $col?: {},
```
Comment (shannonmoeller) [inline src/table-semantic/styled-components.js:52]: @chasestarr Would be happy to have input on how to make these `{}` types more specific.

File: src/table-semantic/__tests__/table-builder.test.js [hunk 2]
```diff
@@ -151,18 +151,91 @@ describe('Table Semantic Builder', () => {
...
+    headCells
       .at(1)
-      .simulate('click');
+      .simulate('focus')
+      .simulate('keydown', {key: 'a'});
+
+    headCells.at(1).simulate('blur');
+
+    expect(mockOnSort.mock.calls.length).toBe(4);
+    expect(mockOnSort.mock.calls).toEqual([['foo'], ['bar'], ['bar'], ['bar']]);
```
Comment (shannonmoeller) [inline src/table-semantic/__tests__/table-builder.test.js:177]: Added these to fill gaps in test coverage.

File: src/table-semantic/styled-components.js (hunk 1 above)
Comment (chasestarr) [reply src/table-semantic/styled-components.js:52]: if there's a generic param that defaults to `{}`, can flow infer when a defined type is provided in `$col`?

Comment (shannonmoeller) [reply src/table-semantic/styled-components.js:52]: I would love to do that, but I can't figure out how to use generics and `styled<>` together.

```jsx
type StyledCompPropsT<T> = {
  col: T,
};

const StyledComp = styled<StyledCompPropsT<T>>('div', () => ({}));
// Flow error: Cannot resolve name `T`.
```

**Comment by curtislin7** [comment]
Hey @chasestarr and @shannonmoeller! Really appreciate the new feature. I linked a related issue that is pertinent #3394.

**Pull request merged**